backend/service/tagging.py

```diff
@@ -0,0 +1,177 @@
+import time
+from threading import Thread
+
+from backend.solr import SolrDocStorage, SolrDoc, SolrDocKeyword, \
+    SolrDocKeywordTypes
+from utils.data_preprocessing import lemmatize_keywords, \
+    create_automated_keywords
+
+
+class TaggingJob:
+    def __init__(self, job_id: str, status='started', progress=0):
+        self.job_id = job_id
+        self.status = status
+        self.progress = progress
+
+
+class KWMJob(Thread, TaggingJob):
+    def __init__(self, keywords: dict, job_id, solr_docs: SolrDocStorage, *doc_ids: str):
+        super().__init__()
+        self.keywords = keywords
+        self.doc_ids = doc_ids
+        self.job_id = job_id
+        self.solr_docs = solr_docs
+        self.status = 'STARTED'
+        self.progress = 0
+        self.cancelled = False
+        self.time_remaining = -1
+
+    def run(self):
+        """
+        Applies a keyword model on every document in Solr.
+        The idea is to search the content in Solr for the lemmatized_keyword if it is found
+        the (normal)keyword and its parents are applied.
+
+        :param keywords: dict of keywords and corresponding parents
+        :param doc_ids:
+        :param job_id
+        :return:
+        """
+        self.status = 'TAGGING_JOB.LEMMA_START'
+        lemmatized_keywords = lemmatize_keywords(self.keywords)
+        self.status = 'TAGGING_JOB.LEMMA_END'
+        self.progress = 20
+
+        self.status = 'TAGGING_JOB.DOC_FIND'
+        id_query = self.solr_docs.build_id_query(self.doc_ids)
+        self.status = 'TAGGING_JOB.DOC_FOUND'
+        changed_docs = {}
+        self.status = 'TAGGING_JOB.APPLY_KWM'
+        start_time = time.time()
+        time_index = 0
+        iteration_time = None
+        idx = 0
+        for lemmatized_keyword, (keyword, parents) in zip(
+            lemmatized_keywords, self.keywords.items()
+        ):
+            if self.cancelled:
+                break
+            if idx == 0:
+                progress_step = 80 / len(lemmatized_keywords)
```

Comment:
magic number

also cant the initialization of `progress_step` be moved out of the loop?